wallet-new/src/Cardano/Wallet/Kernel/ChainState.hs

```diff
@@ -0,0 +1,365 @@
+    , csmSoftwareVersion = upSoftwareVersion
+    , csmScriptVersion   = bvmScriptVersion
+    , csmMaxTxSize       = bvmMaxTxSize
+    })
+  where
+    UnsafeUpdateProposal{..} = cpsUpdateProposal
+    BlockVersionModifier{..} = upBlockVersionMod
+
+-- | Apply a chain state modifier to a chain state
+applyChainStateModifier :: ChainStateModifier -> ChainState -> ChainState
+applyChainStateModifier ChainStateModifier{..} ChainState{..} = ChainState{
+      csBlockVersion    =                           csmBlockVersion
+    , csSoftwareVersion =                           csmSoftwareVersion
+    , csScriptVersion   = fromMaybe csScriptVersion csmScriptVersion
+    , csMaxTxSize       = fromMaybe csMaxTxSize     csmMaxTxSize
+    }
+
+{-------------------------------------------------------------------------------
+  Chain summary
+-------------------------------------------------------------------------------}
+
+-- | Self-consistent summary of the current tip of the chain
+data ChainBrief = ChainBrief {
+      -- | Header hash of the chain tip, according to the node DB
+      cbTip      :: !HeaderHash
+
+      -- | Slot ID of the tip
+      --
+      -- NOTE: This is /not/ the "current" slot ID, but rather the slot
+      -- associated with the 'cbTip'. The intention is that 'ChainBrief'
+      -- provides a consistent view of the node's database for restoration
+      -- purposes. If the current slot ID (based on timestamp) is needed, use
+      -- the 'MonadSlots' interface (but I strongly suspect 'MonadSlots' may in
+      -- fact only be needed by the underlying node).
+    , cbSlotId   :: !SlotId
+
+      -- | Header of the previous main block
+      --
+      -- If there is no previous main block, this is set to the genesis hash.
+    , cbPrevMain :: !HeaderHash
+
+      -- | The chain state at the time of 'cbTip'
+      --
+      -- Implementation note: Although we have no way of verifying that these
+      -- actually match up, the hope is that since we read all these values
+      -- while locking the node state, that they will be consistent with each
+      -- other. That might be overly optimistic.
+    , cbState    :: !ChainState
+    }
+
+-- | Dummy value that can be used as a placeholder
+dummyChainBrief :: ChainBrief
+dummyChainBrief = ChainBrief {
+      cbTip      = unsafeHash ()
+    , cbSlotId   = SlotId (EpochIndex 0) (UnsafeLocalSlotIndex 0)
+    , cbPrevMain = unsafeHash ()
+    , cbState    = ChainState {
+          csBlockVersion    = BlockVersion 0 0 0
+        , csSoftwareVersion = SoftwareVersion (ApplicationName "dummyChainBrief") 0
+        , csScriptVersion   = 0
+        , csMaxTxSize       = fromBytes 65536
+        }
+    }
+
+-- | Get 'ChainBrief' for current chain tip
+--
+-- Returns
+--
+-- * 'Nothing' if the current tip is the genesis block
+-- * The previous block if the current tip is an epoch boundary block (EBB)
+-- * The current block if the current tip is a regular block.
+getChainBrief :: HasCallStack
```

Comment:
I find it incredibly useful when exceptions capture the context in which they were thrown; otherwise it gets caught waaaay later and you have no idea where it came from. By adding the callstack to the exception, if we throw it, we know, "it was thrown on line such-and-such, and that got called from ...".